rust/src/builders/withdrawal_builder.rs

```diff
@@ -0,0 +1,131 @@
+use crate::*;
+use crate::builders::witness_builder::{InputAggregateWitnessData, PartialPlutusWitness};
+
+use super::witness_builder::{RequiredWitnessSet, NativeScriptWitnessInfo, PlutusScriptWitnessInfo};
+
+// comes from witsVKeyNeeded in the Ledger spec
+pub fn withdrawal_required_wits(address: &RewardAddress, required_witnesses: &mut RequiredWitnessSet) -> () {
+    let cred = &address.payment_cred();
+    if let Some(keyhash) = &cred.to_keyhash() {
+        required_witnesses.add_vkey_key_hash(&keyhash);
+    }
+    if let Some(script_hash) = &cred.to_scripthash() {
+        required_witnesses.add_script_hash(&script_hash);
+        // recall: no datum hash for reward withdrawals
+    };
+}
+
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct WithdrawalBuilderResult {
+    address: RewardAddress,
+    amount: Coin,
+    aggregate_witness: Option<InputAggregateWitnessData>,
+    required_wits: RequiredWitnessSet,
+}
+
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct SingleWithdrawalBuilder {
+    address: RewardAddress,
+    amount: Coin,
+}
+
+#[wasm_bindgen]
+impl SingleWithdrawalBuilder {
+    pub fn new(address: &RewardAddress, amount: &Coin) -> Self {
+        Self {
+            address: address.clone(),
+            amount: amount.clone(),
+        }
+    }
+
+    pub fn skip_witness(&self) -> WithdrawalBuilderResult {
+        let mut required_wits = RequiredWitnessSet::default();
+        withdrawal_required_wits(&self.address, &mut required_wits);
+
+        WithdrawalBuilderResult {
+            address: self.address.clone(),
+            amount: self.amount.clone(),
+            aggregate_witness: None,
+            required_wits: required_wits.clone(),
+        }
+    }
+
+    pub fn vkeys(&self, vkeys: &Vkeys) -> Result<WithdrawalBuilderResult, JsError> {
+        let mut required_wits = RequiredWitnessSet::default();
+        withdrawal_required_wits(&self.address, &mut required_wits);
+        let mut required_wits_left = required_wits.clone();
+
+        // the user may have provided more witnesses than required. Strip it down to just the required wits
```

Comment:
Good point. Not sure if there is anything that can be done about this though since the point or this PR is to move away from dummy witnesses since it doesn't work for non-Plutus input types